README.md
# Base-Management-Suite
A collection of tools and programs which provide Minecraft Base Management solutions.


storage_suite/changelog.txt
I Don't Usually do this, so here's hoping.

Storage Version 2.5:
    Fixed minor graphical bugs with version 2. Moved around component calls to give the illusion of speed.
    
Client Version 0.1:
    Got angry at tier 1 GPUs, want to move calls fully to buffer but also don't want to drop tier 1 support.

Storage Version 2.7:
    Added basic network handling, still needs to be reworked
    
Client Version 0.8:
    First fully functional release. Has a few graphical bugs and limitations to be fixed before V1.

    Storage now requires GUI.lua
    


storage_suite/craft.lua
-- COPYRIGHT AUG 2020, James Gottshall --
-- Version: 0.3 --
-- Crafting manager --

local comp = require("component")
local trans = comp.transposer
local sides = require("sides")
local event = require("event")
local modem = comp.modem

-- Constants --
local input_side = sides.south
local output_side = sides.up
local crafter_side = sides.west
local template_side = sides.east
local storage_port = 86

-- Right Up There /\
print("Make Sure to Edit Constants at Top of Program!")

function craft(recipe_name)
	
end

function new_recipe()
	for i=1, 9 do
		trans.get
	end
end

function main()
	modem.open(storage_port)
	while true do
		local id, localNetworkCard, remoteAddress, port, distance, payload = event.pullMultiple("modem_message", "key_down", "interrupted")
		if id == "modem_message" and port == storage_port then
			print("Recieved Message: " .. payload)
		elseif id == "key_down" then
			if string.lower(string.char(remoteAddress)) == "n" then
				new_recipe()	
			end
		elseif id == "interrupted" then
			break
		end
	end
	modem.close(storage_port)
end

main()

storage_suite/client.lua
-- COPYRIGHT AUG 2020, James Gottshall --
-- Version 0.8 --
-- Thin Client for the storage system suite --

local comp = require("component")
local event = require("event")
local gpu = comp.gpu
local serial = require("serialization")
local modem = comp.modem
local storage_port = 86
modem.open(storage_port)
local Swidth, Sheight = gpu.getResolution()

-- Activate Wireless Card --
if modem.isWireless() then
	modem.setStrength(200);
end

-- I hate tier 1 GPUs, they only have buffer space for a single screen
local ADV_GPU = gpu.maxDepth() > 1 -- Whether to use VRAM buffers and colors

local message_bar = ""
local global_options = {}
-- Network Callback --
function network_message(_,_,from_addr,port,dist,...)
	msg = {...}
	if port ~= storage_port then return end
	if msg[1] == "UPDATE" then
		global_options = serial.unserialize(msg[2])
        message_bar = "Ledger Rebuilt"
        event.push("rebuilt")
	elseif msg[1] == "REQUEST_STATUS" then
        message_bar = math.floor(msg[2]) .. " Items Fetched"
        event.push("motion") -- Easy way to display how many items were fetched
	end
end
event.listen("modem_message", network_message)

-- Popup for entering amount. Default is 64. Backspace not supported.
function amount_popup()
	local popup_fg = 0xffffff
	local popup_bg = 0x0000ff
	local oldbg, oldfg
	
	if ADV_GPU then 
		oldbg = gpu.setBackground(popup_bg)
		oldfg = gpu.setForeground(popup_fg)
	end
	gpu.set(Swidth/2-4,Sheight/2,"/--------\\")
	gpu.set(Swidth/2-4,Sheight/2+1,"| Amount |")
	gpu.set(Swidth/2-4,Sheight/2+2,"|        |")
	gpu.set(Swidth/2-4,Sheight/2+3,"\\--------/")
	
	local input = ""
	while true do
		gpu.set(Swidth/2-(string.len(input)/2)+1,Sheight/2+2,input)
		id,_,char,code = event.pullMultiple("key_down","interrupted")
		if id == "interrupted" then return 0 end
		if id == "key_down" then
			if code==28 then
				if ADV_GPU then 
					gpu.setBackground(oldbg)
					gpu.setForeground(oldfg)
				end
				if string.len(input) > 0 then
					return tonumber(input)
				else
					return 64
				end
			end
			if char and tonumber(string.char(char)) then
				input = input .. string.char(char)
			end
		end
	end
end


-- The main part of the request GUI
function request_menu()
	modem.broadcast(storage_port, "GET")
	
	-- GUI Variables
	local cursor = {x=1,y=1} -- Where the selection cursor is. Holy Crap I did this backwards but its too late now
	local scroll = 0
	local loop = true
	-- local Swidth,Sheight = gpu.getResolution()
	local search = ""
	local cursor_color = 0x03fcf4
	local bg_color = 0x2e3b3a
	local options = global_options
	local motionTimer = true
	local buffer = gpu.allocateBuffer(Swidth, Sheight - 2)
    if not buffer then gpu.freeAllBuffers() buffer = gpu.allocateBuffer(Swidth, Sheight -2) end
	
	local function convert(x,y)
		-- Converts 2D coords to 1D, only necessary because I am big dumb. Could probably do this different
		return x*2 - y%2
	end
	
	if ADV_GPU then gpu.setBackground(bg_color) end
	gpu.fill(1,1,Swidth,Sheight," ") -- Clear Screen
	
	-- GUI Loop --
	while loop do
		-- Get User Input Events
		id,addr,char,code,playerName = event.pullMultiple("key_up","key_down","interrupted","rebuilt","motion")
		if id=="interrupted" then 
			gpu.fill(1,1,Swidth,Sheight," ")
            gpu.freeBuffer(buffer)
			modem.close(storage_port)
			return
		end
		if id=="key_down" then
			if code==200 then --Up Arrow
				if cursor.x > 1 then
					cursor.x = cursor.x - 1
				end
				if scroll > 0 and cursor.x - scroll < 2 then
					scroll = scroll - 1
				end
			elseif code==203 then --Left Arrow
				cursor.y = 1
			elseif code==205 then --Right Arrow
				if #options%2==0 or cursor.x*2 - 1 < #options then
					cursor.y = 2
				end
			elseif code==208 then --Down Arrow
				if 2*cursor.x < #options and (convert(cursor.x+1,2) <= #options or cursor.y == 1) then
					cursor.x = cursor.x + 1
				end
				if cursor.x - scroll >= Sheight - 2 then
					scroll = scroll + 1
				end
			elseif code==28 then -- Enter
				local amt = amount_popup()
				message_bar = "Trying to Fetch "..amt.." "..options[convert(cursor.x,cursor.y)][1]
				gpu.fill(1,2,Swidth,Sheight-2, " ")
				modem.broadcast(storage_port, "FETCH", options[convert(cursor.x,cursor.y)][1], tonumber(amt))
				search = ""
			elseif char then
				search = string.char(char)
				for i=1,#options do
					if string.lower(string.sub(options[i][1],1,1)) == string.lower(search) then
						cursor.x = math.ceil(i/2)
						cursor.y = 2 - (i%2)
						scroll = cursor.x - 1
						break
					end
				end
			end
		end
		if id=="rebuilt" then
			options = global_options
		elseif id=="motion" then
			if motionTimer then
				modem.broadcast(storage_port, "GET")
				motionTimer = false
				event.timer(10, function() motionTimer = true end)
			end
		end
		
		if options[convert(cursor.x,cursor.y)] then
			gpu.set(Swidth/2,1,"Amount: ".. math.floor(options[convert(cursor.x,cursor.y)][2]) .."x            ")
		end
		-- Display item list --
		if ADV_GPU then 
            gpu.setActiveBuffer(buffer) 
            gpu.setBackground(bg_color)
        end -- Use Buffers if not on tier 1
        gpu.fill(1,1,Swidth,Sheight-2," ")
		for i=1,2*(Sheight-2) do
			if ADV_GPU and convert(cursor.x - scroll,cursor.y) == i then
				gpu.setBackground(cursor_color)
			elseif ADV_GPU then
				gpu.setBackground(bg_color)
			end
			if not options[i + 2*scroll] then break end 
			local label = options[i + 2*scroll][1]
			label = string.sub(label,1,math.min(#label,Swidth/2 - 3))
			if i%2~=0 then
				gpu.set(1,math.floor(i/2) + 1," ".. label)
			else
				gpu.set(Swidth/2 + 1,i/2," ".. label)
			end
		end
		if ADV_GPU then gpu.setBackground(cursor_color) end
		gpu.set((cursor.y-1)*(Swidth/2)+1,cursor.x - scroll,">")
		if ADV_GPU then 
			gpu.setActiveBuffer(0)
			gpu.bitblt(0,1,2,Swidth,Sheight - 2,buffer)
			gpu.setBackground(bg_color)
		end
        gpu.fill(1,Sheight,Swidth,1, " ")
        gpu.set(Swidth/2 - string.len(message_bar)/2,Sheight, message_bar)
	end
end

request_menu()


storage_suite/storage.lua
-- COPYRIGHT AUG 2020, James Gottshall --
-- Version: 2.7 --

local comp = require("component")
local sides = require("sides")
local event = require("event")
local gpu = comp.gpu
local thread = require("thread")
local color = gpu.maxDepth() > 1
local modem = comp.modem
local storage_port = 86
modem.open(storage_port)
-- Open wireless modem, if it exists
if modem.isWireless() then
	modem.setStrength(200)
end
local serial = require("serialization")
local gui = require("gui")

local transposers = {}
for address, name in comp.list("transposer", true) do
	table.insert(transposers, comp.proxy(address))
end


-- Item Ledger is a table with indexed by the names of items holding table 
-- {size, slots={slot, size, side, transposer}
-- Currently can not tell the difference between certain items such as wool and stained glass
local item_ledger = {} -- Reachable from all functions
local free_space = 0
local global_options = {} -- Formatted indexable version of the ledger
function build_ledger()
	if thread.current() then thread.current():suspend() end
	item_ledger = {}
	free_space = 0
	local sidesList = {sides.north, sides.south, sides.east, sides.west}
	for i,t in ipairs(transposers) do
		for j,side in ipairs(sidesList) do
			local count = 1
			for item in t.getAllStacks(side) do
				if item.name then
					if item_ledger[item.label] then
						item_ledger[item.label].size = item_ledger[item.label].size + item.size
						table.insert(item_ledger[item.label].slots, {slot=count, size=item.size, side=side, transposer=i})
					else
						item_ledger[item.label] = {} 
						item_ledger[item.label].size = item.size
						item_ledger[item.label].slots = {{slot=count, size=item.size, side=side, transposer=i}}
						item_ledger[item.label].label = item.label
					end
				else
					free_space = free_space + 1
				end
				count = count + 1
			end
		end
	end
	-- Options table entry looks like this {label, amount} --
	global_options = {}
	for name, v in pairs(item_ledger) do
		table.insert(global_options, {name,v.size})
	end
	table.sort(global_options, function(a,b) return a[1] < b[1] end)
	event.push("rebuilt")
    modem.broadcast(storage_port, "UPDATE", serial.serialize(global_options))
	return true, global_options
end

-- Popup for entering amount. Default is 64. Backspace not supported.
function amount_popup()
	local width,height = gpu.getResolution()
	local popup_fg = 0xffffff
	local popup_bg = 0x0000ff
	local oldbg, oldfg
	
	if color then 
		oldbg = gpu.setBackground(popup_bg)
		oldfg = gpu.setForeground(popup_fg)
	end
	gpu.set(width/2-4,height/2,"/--------\\")
	gpu.set(width/2-4,height/2+1,"| Amount |")
	gpu.set(width/2-4,height/2+2,"|        |")
	gpu.set(width/2-4,height/2+3,"\\--------/")
	
	local input = ""
	while true do
		gpu.set(width/2-(string.len(input)/2)+1,height/2+2,input)
		id,_,char,code = event.pullMultiple("key_down","interrupted")
		if id == "interrupted" then return 0 end
		if id == "key_down" then
			if code==28 then
				if color then 
					gpu.setBackground(oldbg)
					gpu.setForeground(oldfg)
				end
				if string.len(input) > 0 then
					return tonumber(input)
				else
					return 64
				end
			end
			if char and tonumber(string.char(char)) then
				input = input .. string.char(char)
			end
		end
	end
end

-- The main part of the request GUI
function request_menu(ledger)
	-- GUI Variables
	local cursor = {x=1,y=1} -- Where the selection cursor is. Holy Crap I did this backwards but its too late now
	local scroll = 0
	local loop = true
	local screenX,screenY = gpu.getResolution()
	local search = ""
	local cursor_color = 0xff0000
	local bg_color = 0x004455
	local build_thread = nil
	local request_thread = nil
	local options = ledger
	local motionTimer = true
	
	local function convert(x,y)
		-- Converts 2D coords to 1D, only necessary because I am big dumb. Could probably do this different
		return x*2 - y%2
	end
	
	if color then gpu.setBackground(bg_color) end
	gpu.fill(1,1,screenX,screenY," ") -- Clear Screen
	
	-- GUI Loop --
	while loop do
		-- Get User Input Events
		id,addr,char,code,playerName = event.pullMultiple("key_up","key_down","interrupted","rebuilt","motion")
		if id=="interrupted" then 
			gpu.fill(1,1,screenX,screenY," ")
			modem.close(storage_port)
			return
		end
		if id=="key_down" then
			if code==200 then --Up Arrow
				if cursor.x > 1 then
					cursor.x = cursor.x - 1
				end
				if scroll > 0 and cursor.x - scroll < 2 then
					scroll = scroll - 1
					gpu.fill(1,2,screenX,screenY," ")
				end
			elseif code==203 then --Left Arrow
				cursor.y = 1
			elseif code==205 then --Right Arrow
				if #options%2==0 or cursor.x*2 - 1 < #options then
					cursor.y = 2
				end
			elseif code==208 then --Down Arrow
				if 2*cursor.x < #options and (convert(cursor.x+1,2) <= #options or cursor.y == 1) then
					cursor.x = cursor.x + 1
				end
				if cursor.x - scroll >= screenY - 2 then
					scroll = scroll + 1
					gpu.fill(1,2,screenX,screenY," ")
				end
			elseif code==28 then -- Enter
				local amt = amount_popup()				
				gpu.fill(1,2,screenX,screenY-2, " ")
				request_thread = thread.create(fill_request, options[convert(cursor.x,cursor.y)][1], amt)
				build_thread = thread.create(build_ledger)
				search = ""
			elseif char then
				search = string.char(char)
				for i=1,#options do
					if string.lower(string.sub(options[i][1],1,1)) == string.lower(search) then
						cursor.x = math.ceil(i/2)
						cursor.y = 2 - (i%2)
						scroll = cursor.x - 1
						break
					end
				end
			end
		end
		if id=="rebuilt" then
			options = global_options
		elseif id=="motion" then
			if motionTimer then 
				gpu.fill(1,1,screenX,screenY," ")
				motionTimer = false
				build_thread = thread.create(build_ledger)
				event.timer(10, function() motionTimer = true end)
			end
		end
		
		gpu.fill(1,1,screenX,1," ")
		gpu.set(1,1,"Free Slots: "..free_space)
		-- gpu.set(screenX/2,1,"Key: "..char.." Code: "..code)
		if options ~= global_options then
			options = global_options
		end
		if options[convert(cursor.x,cursor.y)][2] then
			gpu.set(screenX/2 + 1,1,"Amount: "..math.floor(options[convert(cursor.x,cursor.y)][2]).."x")
		end
		-- gpu.set(1,screenY,"X: "..cursor.x.." Y: "..cursor.y) -- For testing cursor position
		-- Display item list
		for i=1,2*(screenY-2) do
			if color and convert(cursor.x - scroll,cursor.y) == i then
				gpu.setBackground(cursor_color)
			elseif color then
				gpu.setBackground(bg_color)
			end
			if not options[i + 2*scroll] then break end 
			local label = options[i + 2*scroll][1]
			label = string.sub(label,1,math.min(#label,screenX/2 - 3))
			label = gui.rpad(label, screenX / 2 - 1)
			if i%2~=0 then
				gpu.set(1,math.floor(i/2)+2," ".. label)
			else
				gpu.set(screenX/2 + 1,i/2+1," ".. label)
			end
		end
		if color then gpu.setBackground(cursor_color) end
		gpu.set((cursor.y-1)*(screenX/2)+1,cursor.x+1-scroll,">")
		if color then gpu.setBackground(bg_color) end
		
		-- This solution seems to be marginally faster, allows graphics to display before calling a ton of yield calls
		if request_thread then
			request_thread:resume()
			request_thread = nil
		end
		if build_thread then
			build_thread:resume()
			build_thread = nil
		end
	end
end

function empty_slot(trans, side)
	local count = 1
	for i in trans.getAllStacks(side) do
		if i.name == nil then return count end
		count = count + 1
	end
end

-- Grabs requested item, returns amount fetched
function fill_request(item_name, amount)
	if thread.current() then thread.current():suspend() end
	local item = item_ledger[item_name] -- Ledger info about the item, see build_ledger
	if amount > item.size then
		amount = item.size
	end
	local total = 0
	while amount > 0 do
		local slot = item.slots[#item.slots]
		local empty = empty_slot(transposers[slot.transposer], sides.up)
		local transfer = transposers[slot.transposer].transferItem(slot.side, sides.up, math.min(slot.size, amount), slot.slot, empty)
		-- TODO actually update ledger
		amount = amount - transfer
		total = total + transfer
		slot.size = slot.size - transfer
		if slot.size <= 0 then table.remove(item.slots, #item.slots) end
	end
	return total
end

-- Network Callback Function
function network_request(_,_,from_addr,port,dist,...)
	local msg = {...}
	if port ~= storage_port then return end
	if msg[1] == "GET" then
		-- msg[2] is the start index and msg[3] is the end index
		-- TODO: implement above
		modem.broadcast(storage_port, "UPDATE", serial.serialize(global_options))
	elseif msg[1] == "FETCH" then
		-- msg[2] is the item name and msg[3] is the amount
		modem.send(from_addr, port, "REQUEST_STATUS", fill_request(msg[2], tonumber(msg[3])))
	end
end
event.listen("modem_message", network_request)

-- Program Starts Here --
print("Building Ledger\n")
local _,ledger = build_ledger()
print("Free Slots Remaining: "..free_space)
request_menu(ledger)
